src/testRunner/unittests/semver.ts

```diff
@@ -120,129 +98,826 @@ namespace ts {
+
+            function assertVersionRange(version: string, good: string[], bad: string[]): () => void {
+                return () => {
+                    const range = VersionRange.tryParse(version)!;
+                    assert(range);
+                    for (const g of good) {
+                        assert.isTrue(range.test(g), g);
+                    }
+                    for (const b of bad) {
+                        assert.isFalse(range.test(b), b);
+                    }
+                };
+            }
+
+            it("< works", assertVersionRange("<3.8.0", ["3.6", "3.7"], ["3.8", "3.9", "4.0"]));
+            it("<= works", assertVersionRange("<=3.8.0", ["3.6", "3.7", "3.8"], ["3.9", "4.0"]));
+            it("> works", assertVersionRange(">3.8.0", ["3.9", "4.0"], ["3.6", "3.7", "3.8"]));
+            it(">= works", assertVersionRange(">=3.8.0", ["3.8", "3.9", "4.0"], ["3.6", "3.7"]));
+
+            it("< works with prerelease", assertVersionRange("<3.8.0-0", ["3.6", "3.7"], ["3.8", "3.9", "4.0"]));
+            it("<= works with prerelease", assertVersionRange("<=3.8.0-0", ["3.6", "3.7"], ["3.8", "3.9", "4.0"]));
+            it("> works with prerelease", assertVersionRange(">3.8.0-0", ["3.8", "3.9", "4.0"], ["3.6", "3.7"]));
+            it(">= works with prerelease", assertVersionRange(">=3.8.0-0", ["3.8", "3.9", "4.0"], ["3.6", "3.7"]));
+
+            function assertRange(rangeText: string, versionText: string, inRange: boolean) {
                 const range = new VersionRange(rangeText);
                 const version = new Version(versionText);
                 assert.strictEqual(range.test(version), inRange, `Expected version '${version}' ${inRange ? `to be` : `to not be`} in range '${rangeText}' (${range})`);
             }
+
             theory("comparators", assertRange, [
-                ["", "1.0.0"],
-                ["*", "1.0.0"],
-                ["1", "1.0.0"],
+                // empty (matches everything)
+                ["", "2.0.0", true],
+                ["", "2.0.0-0", true],
+                ["", "1.1.0", true],
+                ["", "1.1.0-0", true],
+                ["", "1.0.1", true],
+                ["", "1.0.1-0", true],
+                ["", "1.0.0", true],
+                ["", "1.0.0-0", true],
+                ["", "0.0.0", true],
+                ["", "0.0.0-0", true],
+
+                // wildcard major (matches everything)
+                ["*", "2.0.0", true],
+                ["*", "2.0.0-0", true],
+                ["*", "1.1.0", true],
+                ["*", "1.1.0-0", true],
+                ["*", "1.0.1", true],
+                ["*", "1.0.1-0", true],
+                ["*", "1.0.0", true],
+                ["*", "1.0.0-0", true],
+                ["*", "0.0.0", true],
+                ["*", "0.0.0-0", true],
+
+                // wildcard minor
                 ["1", "2.0.0", false],
-                ["1.0", "1.0.0"],
+                ["1", "2.0.0-0", false], // failing

```

Comment:
what does `// failing` mean here?

Reply:
Nothing important. I've removed it, thanks.